File: Core-backend-layer/Programming-fundamentals/Python/DesignPatterns/src/memento/reservation.py
from memento.reservation_memento import ReservationMemento

class Reservation:
    def __init__(self, customer_name, party_size, date_time):
        self.customer_name = customer_name
        self.party_size = party_size
        self.date_time = date_time

    def get_customer_name(self):
        return self.customer_name

    def get_party_size(self):
        return self.party_size

    def get_date_time(self):
        return self.date_time

    # Setter method to update the reservation details
    def update_reservation(self, customer_name, party_size, date_time):
        self.customer_name = customer_name
        self.party_size = party_size
        self.date_time = date_time

    # Create a memento to save the current state
    def save(self):
        return ReservationMemento(self.customer_name, self.party_size, self.date_time)

    # Restore the reservation state from a memento
    def restore(self, memento):
        self.customer_name = memento.get_customer_name()
        self.party_size = memento.get_party_size()
        self.date_time = memento.get_date_time()

File: Core-backend-layer/Programming-fundamentals/Java/DesignPatterns/src/memento/Reservation.java
package memento;

public class Reservation {
    private String customerName;
    private int partySize;
    private String dateTime;

    public Reservation(String customerName, int partySize, String dateTime) {
        this.customerName = customerName;
        this.partySize = partySize;
        this.dateTime = dateTime;
    }

    public String getCustomerName() {
        return customerName;
    }

    public int getPartySize() {
        return partySize;
    }

    public String getDateTime() {
        return dateTime;
    }

    // Setter method to update the reservation details
    public void updateReservation(String customerName, int partySize, String dateTime) {
        this.customerName = customerName;
        this.partySize = partySize;
        this.dateTime = dateTime;
    }

    // Create a memento to save the current state
    public ReservationMemento save() {
        return new ReservationMemento(customerName, partySize, dateTime);
    }

    // Restore the reservation state from a memento
    public void restore(ReservationMemento memento) {
        this.customerName = memento.getCustomerName();
        this.partySize = memento.getPartySize();
        this.dateTime = memento.getDateTime();
    }
}